064-topic-replication-factor-change.md

```diff
@@ -0,0 +1,291 @@
+
+This is a non exhaustive list of use cases that will become significantly easier:
+
+- Set a lower replication factor for non-critical topics to reduce the infrastructure costs (compute, storage and network).
+- Set a lower replication factor to deal with resource shortage (trade-off between fault-tolerance and resource utilization).
+- Set a higher replication factor for critical topics to improve durability and fault tolerance.
+
+A tighter integration with Cruise Control is also one of the long term goals of the Strimzi project. 
+
+## Proposal
+
+The replication factor change feature can be implemented by integrating the Topic Operator with Cruise Control.
+The Cluster Operator is already integrated with Cruise Control, so we can factor out the existing logic and classes.
+
+Cruise Control creates a task for each operation, such as topic configuration and cluster rebalance.
+A task is first created in the Active state, then moves to InExecution, and finally to Completed or CompletedWithError.
+At most `max.active.user.tasks` tasks can be Active at the same time (5 by default), and only one task can be InExecution at any given time.
+During the execution, a task is divided in one or more batches, which are then executed one after the other.
+A task may include multiple topics and it is marked as completed once the whole task's status is Completed.
+
+### Topic configuration
+
+The [`topic_configuration`](https://github.com/linkedin/cruise-control/wiki/REST-APIs#change-kafka-topic-configuration) endpoint is used to request a topic configuration update, and currently only supports replication factor changes.
+Once accepted, the replication factor change is put into the Cruise Control's task queue and executed asynchronously.
+
+The two required parameters for this endpoint are `topic` (regular expression to specify subject topics) and `replication_factor` (target replication factor).
+In order to group multiple changes in the same request to Cruise Control, the Topic Operator uses a JSON payload containing topics grouped by their replication factor.
+
+This is useful with bulk replicas change, which can happen in some edge cases.
+A potentially big backlog of messages may be created when the Topic Operator is not running, or there is an issue with Cruise Control or Kafka.
+Another example is a GitOps pipeline delivering many replicas changes at the same time to reduce the storage occupation of non-critical topics.
+
+```sh
+$ curl -vv -X POST -H "Content-Type: application/json" -d '{replication_factor:{topic_by_replication_factor:{2:topic1|topic3|topic4,3:topic2|topic5}}}' \
+  "http://localhost:9090/kafkacruisecontrol/topic_configuration?skip_rack_awareness_check=true&dryrun=false&json=true" | jq
+> POST /kafkacruisecontrol/topic_configuration?skip_rack_awareness_check=true&dryrun=false&json=true HTTP/1.1
```

Comment:
it's not clear to me how this example maps to the endpoint definition.
The REST endpoint is missing the `topic` and `replication_factor` parameters which are defined as required (in the proposal itself).
There is also a JSON passed as body with some topics and replication factor information but AFAICS the Cruise Control Wiki doesn't mention it.

Reply:
The two required parameters are contained in the JSON payload, which is described here: https://github.com/linkedin/cruise-control/wiki/Change-topic-replication-factor-through-Cruise-Control#instruction

Reply:
But this link is not in the proposal, or? You just posted the REST API Wiki one where the JSON payload is not explained at all. And I think it's mostly fault of the Cruise Control community having a dedicated page for the change replication factor stuff with additional instruction. So I think the link you just shared should be in the proposal.

Reply:
Ok, link added.